### DISCUSSION TITLE
How to do vertical alignment

### DISCUSSION THREAD

**Discussion by Puge96** [QuestPDF/QuestPDF#649] [Q&A] [answered]
Hi!

I'm having troubles with vertical alignment of items.

```
public void Compose(IDocumentContainer container)
{
    container
        .Page(page =>
        {
            page.Size(PageSizes.A4.Landscape());
            page.DefaultTextStyle(style => style.FontSize(10).FontFamily("Arial").FontColor("#3b4151"));
            page.Content().Element(ComposeContent);
        });
}

internal void ComposeContent(IContainer container)
{
    container.PaddingLeft(40f).PaddingTop(40f).PaddingRight(40f).Column(column =>
    {
        column.Item().Border(1).Height(200).Table(table =>
        {
            table.ColumnsDefinition(x =>
            {
                x.RelativeColumn(100);
                x.RelativeColumn(100);
                x.RelativeColumn(100);
            });
            
            table.Cell().Border(1).AlignCenter().AlignTop().Text("Align top");
            table.Cell().Border(1).AlignCenter().AlignMiddle().Text("Align middle");
            table.Cell().Border(1).AlignCenter().AlignBottom().Text("Align bottom");
        });
    });
}
```
Result: 
<img width="841" alt="TableAlignment" src="https://github.com/QuestPDF/QuestPDF/assets/18257257/644eaa30-be1b-480e-bb3c-c727d755762f">

Goal:
<img width="840" alt="tableAlignmentGoal" src="https://github.com/QuestPDF/QuestPDF/assets/18257257/6b1d33d1-f480-4cdd-95ef-8eef9e7093e6">

In this example I wish to align cells within a table but I have the same issue when working with Column layouts; that I cannot get AlignMiddle() & AlignBottom() to do anything.

What am I misunderstanding / doing wrong?

Please advice, thanks.

**Comment by MarcinZiabek** [comment, marked as answer]
The table layout structure takes minimal vertical space as possible. To achieve your requirement, please use the `Row` element. 

```csharp
.Height(200).Border(1).Row(row =>
{
    row.Spacing(50);

    row.RelativeItem().ExtendVertical().AlignTop().Height(50).Background(Placeholders.BackgroundColor());
    row.RelativeItem().ExtendVertical().AlignMiddle().Height(50).Background(Placeholders.BackgroundColor());
    row.RelativeItem().ExtendVertical().AlignBottom().Height(50).Background(Placeholders.BackgroundColor());
});
```

<img width="579" alt="image" src="https://github.com/QuestPDF/QuestPDF/assets/9263853/4a5de265-b10b-4327-9ef7-29245f302b0b">

**Reply by Puge96** [reply]
Sweet! Yet again, thank you very much.

**Reply by MarcinZiabek** [reply]
I decided to improve the `Row` element to respect the provided vertical space, to better match the expected behavior. This change is done within the following commit: https://github.com/QuestPDF/QuestPDF/commit/e0dc65d46131e9d445f7c4706a933b399c858b11

As a result, the following code will be sufficient to achieve the desired layout (please notice the lack of `ExtendVertical` calls):

```csharp
.Height(200).Border(1).Row(row =>
{
    row.Spacing(50);

    row.RelativeItem().AlignTop().Height(50).Background(Placeholders.BackgroundColor());
    row.RelativeItem().AlignMiddle().Height(50).Background(Placeholders.BackgroundColor());
    row.RelativeItem().AlignBottom().Height(50).Background(Placeholders.BackgroundColor());
});
```

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
